module/zfs/dsl_bookmark.c

```diff
@@ -413,6 +420,194 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
+	zfs_bookmark_phys_t bmark_phys = { 0 };
+
+	/* Verify new and target bookmark strings pass bookmark namecheck */
+	{
+		char *target;
+		if (bookmark_namecheck(nvpair_name(pair), NULL, NULL) != 0)
+			return (SET_ERROR(EINVAL));
+		target = fnvpair_value_string(pair);
+		if (target == NULL)
+			return (SET_ERROR(EINVAL));
+		if (bookmark_namecheck(target, NULL, NULL) != 0)
+			return (SET_ERROR(EINVAL));
+	}
+
+	/* Verify that new and target bookmark are on the same dataset */
+	{
+		char *new, *target;
+		char *new_pound, *target_pound;
+		size_t new_ds_len, target_ds_len;
+
+		new = nvpair_name(pair);
+		target = fnvpair_value_string(pair);
+		ASSERT(target != NULL);
+
+		new_pound = strchr(new, '#');
+		ASSERT(new_pound != NULL);
+		new_ds_len = new_pound - nvpair_name(pair);
+
+		target_pound = strchr(target, '#');
+		ASSERT(target_pound != NULL);
+		target_ds_len = target_pound - target;
+
+		if (new_ds_len != target_ds_len)
+			return (SET_ERROR(EINVAL));
+		if (strncmp(new, target, new_ds_len) != 0)
+			return (SET_ERROR(EINVAL));
+	}
+
+	/* Verify that the new bookmark does not already exist */
+	error = dsl_bookmark_lookup(dp, nvpair_name(pair), NULL, &bmark_phys);
+	switch (error) {
+	case 0:
+		return (SET_ERROR(EEXIST));
+	case ESRCH:
+		/* happy path: new bmark doesn't exist, proceed after switch */
+		break;
+	default:
+		return (SET_ERROR(error));
+	}
+
+	/* Verify that the target bookmark exists */
+	error = dsl_bookmark_lookup(dp, fnvpair_value_string(pair),
+	    NULL, &bmark_phys);
+	if (error != 0) {
+		return (SET_ERROR(error));
+	}
+
+	return (0);
+}
+
+static int
+dsl_bookmark_clone_check(void *arg, dmu_tx_t *tx)
+{
+	dprintf("dsl_bookmark_clone_check began\n");
+	dsl_bookmark_clone_arg_t *dbcc = arg;
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	int rv = 0;
+
+	if (!spa_feature_is_enabled(dp->dp_spa, SPA_FEATURE_BOOKMARKS))
+		return (SET_ERROR(ENOTSUP));
+
+	for (nvpair_t *pair = nvlist_next_nvpair(dbcc->dbcc_bmarks, NULL);
+	    pair != NULL; pair = nvlist_next_nvpair(dbcc->dbcc_bmarks, pair)) {
+		int error;
+
+		/*
+		 * We assume that dbcc_bmarks has unique names (keys)
+		 * This is enforced by zfs_ioc_bookmark_clone
+		 */
+
+		error = dsl_bookmark_clone_check_impl(pair,  tx);
+		if (error != 0) {
+			fnvlist_add_int32(dbcc->dbcc_errors, nvpair_name(pair),
+			    error);
+			rv = error;
+		}
+	}
+
+	return (rv);
+}
+
+static void
+dsl_bookmark_clone_sync_impl(const char *new_name, const char *target_name,
+    dmu_tx_t *tx)
+{
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	dsl_dataset_t *bmark_fs_target, *bmark_fs_new;
+	char *target_shortname, *new_shortname;
+	zfs_bookmark_phys_t target_phys;
+	boolean_t target_redacted;
+
+	VERIFY0(dsl_bookmark_hold_ds(dp, target_name, &bmark_fs_target, FTAG,
+	    &target_shortname));
+	VERIFY0(dsl_bookmark_hold_ds(dp, new_name, &bmark_fs_new, FTAG,
+	    &new_shortname));
+	/* Bookmarks must be on the same dataset */
+	VERIFY3U(bmark_fs_target->ds_objset, ==, bmark_fs_new->ds_objset);
```

Comment:
Is this a legitimate and correct way to assert identity of two datasets? Didn't find a `dsl_dataset_equals` function or similar.

Reply:
hat should be `ds_object`, but the question hasn't changed

Reply:
Verifying that they have the same ds_object will work.